copy/multiple.go

```diff
@@ -27,23 +32,90 @@ const (
+
+// Converts imgspecv1.Platform to a comparable format.
+func platformV1ToPlatformComparable(platform imgspecv1.Platform) platformComparable {
+	return platformComparable{Architecture: platform.Architecture,
+		OS:         platform.OS,
+		OSFeatures: strings.Join(platform.OSFeatures, ","),
+		OSVersion:  platform.OSVersion,
+		Variant:    platform.Variant,
+	}
+}
+
+// platformCompresionPair prepares a mapping of platformComparable -> CompressionAlgorithmNames for given digests
+func platformCompresionPair(list internalManifest.List, instanceDigests []digest.Digest) (map[platformComparable]*set.Set[string], error) {
+	platformCompressionMap := make(map[platformComparable]*set.Set[string])
+	for _, instanceDigest := range instanceDigests {
+		instanceDetails, err := list.Instance(instanceDigest)
+		if err != nil {
+			return nil, fmt.Errorf("getting details for instance %s: %w", instanceDigest, err)
+		}
+		platform := platformV1ToPlatformComparable(*instanceDetails.ReadOnly.Platform)
+		_, valueExists := platformCompressionMap[platform]
+		if !valueExists {
+			platformCompressionMap[platform] = set.New[string]()
+		}
+		set := platformCompressionMap[platform]
+		set.AddSlice(instanceDetails.ReadOnly.CompressionAlgorithmNames)
+	}
+	return platformCompressionMap, nil
 }
 
 // prepareInstanceCopies prepares a list of instances which needs to copied to the manifest list.
-func prepareInstanceCopies(instanceDigests []digest.Digest, options *Options) []instanceCopy {
+func prepareInstanceCopies(list internalManifest.List, instanceDigests []digest.Digest, options *Options) ([]instanceCopy, error) {
 	res := []instanceCopy{}
+	platformCompressionMap, err := platformCompresionPair(list, instanceDigests)
+	if err != nil {
+		return res, err
+	}
 	for i, instanceDigest := range instanceDigests {
 		if options.ImageListSelection == CopySpecificImages &&
 			!slices.Contains(options.Instances, instanceDigest) {
 			logrus.Debugf("Skipping instance %s (%d/%d)", instanceDigest, i+1, len(instanceDigests))
 			continue
 		}
+		instanceDetails, err := list.Instance(instanceDigest)
+		if err != nil {
+			return res, fmt.Errorf("getting details for instance %s: %w", instanceDigest, err)
+		}
+		for _, compressionVariant := range options.EnsureCompressionVariantsExist {
+			platform := platformV1ToPlatformComparable(*instanceDetails.ReadOnly.Platform)
+			compressionList := platformCompressionMap[platform]
+			if !slices.Contains(compressionList.Values(), compressionVariant.Algorithm.Name()) {
+				res = append(res, instanceCopy{
+					op:                            instanceCopyClone,
+					sourceDigest:                  instanceDigest,
+					compressionVariant:            compressionVariant.Algorithm,
+					cloneAddPlatform:              instanceDetails.ReadOnly.Platform,
+					cloneAddAnnotations:           maps.Clone(instanceDetails.ReadOnly.Annotations),
+					cloneAddCompressionAlgorithms: []compressionTypes.Algorithm{*compressionVariant.Algorithm},
+				})
```

Comment:
Maybe add the clone’s algorithm to `compressionList` so that we don’t create two clones on duplicate input

Reply:
Yes, this is necessary — not just if `Ensure…Exist` contains the same algorithm twice, but also when the original image already contains two instances for the same platform. (Consider some future algo3, an image with gzip+zstd, and a call with `Ensure…Exist = [algo3]`.

---

An alternative implementation approach would be to not implement `EnsureCompressionVariantsExist` here per-instance, but to do it per-platform:
```go
for i, instanceDigest = instanceDigests { /* … issue instanceCopyCopy only */}
for platform, algos := range platformCompressionMap {
   for _, requiredVariant := range Ensure…Exist {
       if !algos.Contains(requiredVariant) {
          /* issue instanceCopyClone */
       } 
   }
}
```
OTOH that would require us to also build some `comparablePlatform -> (instance digest , OCI platform)` data, so that’s probably much more work.

Just something to keep in mind when looking at the `CopySpecificInstances` interaction.

Reply:
> Maybe add the clone’s algorithm to compressionList so that we don’t create two clones on duplicate input

Did this.

> Just something to keep in mind when looking at the CopySpecificInstances interaction.

For now erroring out if `CopySpecificInstance` is selected and `EnsureCompressionVariantsExist` is set, added a test to verify the same.